### PR TITLE
Make FileAndResourceDirectives work in OSGi environments

### PR DESCRIPTION
**Pull request by aruediger**
These have different URI schemes for resources.

@briantopping Could you give it a spin, please?

### CODE REVIEW

**Comment by johanandren** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Looks reasonable but would ask our local OSGi expert @briantopping to give it a spin :)

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileAndResourceDirectives.scala [hunk 1]
```diff
@@ -222,7 +222,18 @@ object FileAndResourceDirectives extends FileAndResourceDirectives {
...
+        // In OSGi trailing slashes refer to directories. If url already ended with a slash, we wouldn't get here.
+        def isBundleDir = classLoader.getResource(url.getPath + "/") != null
+        if (!proto.equals("bundle") || !isBundleDir) {
+          val conn = url.openConnection()
+          try {
+            conn.setUseCaches(false) // otherwise the JDK will keep the connection open when we close
+            val len = conn.getContentLength
+            val lm = conn.getLastModified
+            Some(ResourceFile(url, len, lm))
+          } finally conn.getInputStream.close() // conn might have already opened the stream
```
Comment (2beaucoup) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileAndResourceDirectives.scala]: Thx @sirthias and @jrudolph for unraveling this mystery. :)

This was originally added for `FileURLConnection`s but who knows how other `UrlConnection`s behave.

**Comment by 2beaucoup** [comment]
`SubstreamSubscriptionTimeoutSpec` failure seems unrelated.

**Comment by drewhk** [comment]
LGTM

**Comment by briantopping** [comment]
Hi guys, I just saw this. I guess I check my email in the morning when my eyes are barely working. I thought email about this was referring to #19331 and didn't see a back-reference to this, so didn't realize the work was being twice.

Is there a reference somewhere for "In OSGi trailing slashes refer to directories"? I've never heard of this, seems like it might be container-specific since my memory is bundle URLs are meant to be opaque.

Anyway, I will build and check later today,

**Comment by rkuhn** [comment]
closed by github due to deletion of the release-2.3-dev branch—please open a new PR against master

**Comment by 2beaucoup** [comment]
reopened against master: #19409

**Pull request closed without merging**